// JS/text_processing_for_typing.js
const puppeteer = require('puppeteer');
const fs = require('fs');

let textForTyping;
let text;
let texts = [];
const STARTNUMBER = 5;
let textNumber = 0;

async function getTextForTyping(res, url) {
  // делаем цикл for и проходимся по тексатм и корректируем как нам нужно
  const browser = await puppeteer.launch({
    headless: 'new',
  });
  const page = await browser.newPage();
  // await page.setViewport({ width: 1080, height: 1024 });

  let source = 'https://ilibrary.ru';

  await page.goto(source);
  let textLink = await page.$eval('.ltst_l > li > a', (elem) =>
    elem.getAttribute('href')
  );
  console.log(textLink);

  await page.goto(`${source}${textLink}`);
  textForTyping = await page.$eval(
    `.p:nth-child(${STARTNUMBER + textNumber})`,
    (elem) => elem.textContent
  );

  await browser.close();

  console.log(textPreparation(textForTyping));

  text = {
    id: textNumber++,
    text: textPreparation(textForTyping),
  };

  texts.push(text);

  postText(res, url, texts);
}

async function postText(res, url, texts) {
  fs.writeFile(`.${url}`, JSON.stringify(texts), (error, data) => {
    if (error) {
      console.log(`ошибка при записи текста в ${url}`);
      return;
    }

    console.log(`текст записан в ${url}`);
    fs.readFile(`.${url}`, { flag: 'r' }, (error, data) => {
      if (error) {
        console.log(`ошибка при чтении файла .${url}`);
        res.end();
        return;
      }
      res.end(data);
      return;
    });
  });
}

function textPreparation(textForTyping) {
  textForTyping = textForTyping.replaceAll('—', '-');

  textForTyping = textForTyping.trim();

  return textForTyping;
}

exports.getTextForTyping = getTextForTyping;
